### PR TITLE
Make preface usable with esy packaging

### PR DESCRIPTION
**Pull request by oteku**
I propose to add a `esy.json` manifest to make this lib usable with esy.sh packaging while it's not published to opam

### CODE REVIEW

**Comment by gr-im** [comment]
Hi @oteku!
Firstly, thank you very much for your contribution. To be very honest, I don't know `esy' at all so it is possible that some of my questions are perfectly stupid.
- Could you summarize the interests of Esy in relation to OPAM, not knowing the project at all, I think that would allow me to better understand the issues of this patch.
- What blocks `esy` from using a package with the `pin` strategy, conventionally used with OPAM?
- I find the diff quite impressive, it adds a lot of files (only in the lock directory) which I find quite annoying. Wouldn't it be better at this point to "vendore" at the user level rather than add an `esy` description?

If @xvw and @d-plaindoux  agree, my idea of the roadmap is to release Preface on opam (in 1.0.0, or whatever) as soon as `Update` is finished. So my view (which is my own) would be to wait for the publication rather than to merge this branch. What do you think?

File: README.md [hunk 1]
```diff
@@ -56,6 +58,24 @@ pin-depends: [
...
+The library can also be installed with [esy](esy.sh) using a resolution in your `package.json` file :
+
+```json
+...
+    "dependencies": {
+      ...
+      "@opam/preface":"*"
+    },
+    "resolutions": {
+        "@opam/preface":"xvw/preface#e02e1bb"
```
[the remark this answers is not in the document]
Comment (oteku) [reply README.md:72]: > Could you summarize the interests of Esy in relation to OPAM, not knowing the project at all, I think that would allow me to better understand the issues of this patch.

esy provide a single CLI to manage scripts (like make) and dependencies with sandboxing from opam, npm or any git repository. I made a more detailed post here https://oteku.github.io/esy-reasonml/

esy sandboxing provide locks (that where not provided by stable opam switch yet) and a better sandboxing strategy (opiniated) https://esy.sh/docs/en/how-it-works.html

Whatever esy is used by many OCaml software developer and it may be good to give them access to Preface whatever is your own preference (cost is just to maintain the esy.lock until preface is released to opam and if your are ok i can do it)

> What blocks esy from using a package with the pin strategy, conventionally used with OPAM?

esy don't rely on local opam installation.
All you need is npm to use esy.

> I find the diff quite impressive, it adds a lot of files (only in the lock directory) which I find quite annoying. Wouldn't it be better at this point to "vendore" at the user level rather than add an esy description?

It just add one file (esy.json). `esy.lock` folder is generated, it's a best practice to commit locks when you want to deliver reproducible builds. But if it bother you i can remove this folder from the distribution.
I contributor are not using esy it will not be a big deal (like lwt)

> Why not using, as an user, this kind of resolution "@opam/preface": "xvw/preface#4cc943486712e9a54f345bf66aeefb0c89c60ab2" which seems iso to opam pin (in https://esy.sh/docs/en/using-repo-sources-workflow.html#with-opam-packages). I'm not pretty sure about what they will say in The only difference is that one should also specify an opam package name

both are doing exacytly the same, I find the use of resolution more convenient cause user of preface will only  to remove resolution when you release preface to opam. 

> So my view (which is my own) would be to wait for the publication rather than to merge this branch. What do you think?
Up to you :) 
I can work on my fork until v1.0

[the remark this answers is not in the document]
Comment (gr-im) [reply README.md:72]: Without the lock folder (even it seems better). I don't have objection about having an esy.json.

Comment (gr-im) [inline README.md:72]: Could you remove the commit avec put `<commit>` in order to make the README not so "obseletable". Since there is a link just below, I think it is not necessary to put the commit in the snipper.

**Comment by gr-im** [comment]
And in addition to modifying the code snippet, could you add a GitHub action that reads the git directory, retrieves the latest commit and attempts an install? Thanks in advance.

**Review by gr-im — APPROVED** [review]
Thanks a lot!
